client/src/panels/WebviewManager.ts

```diff
@@ -25,11 +30,71 @@ export abstract class WebView {
   private _disposables: Disposable[] = [];
   private _onDispose: () => void;
 
+  public constructor(
+    protected readonly extensionUri: Uri,
+    protected readonly title: string,
+  ) {}
+
   set onDispose(disposeCallback: () => void) {
     this._onDispose = disposeCallback;
   }
 
-  abstract render(): WebView;
+  abstract body(): string;
+  abstract l10nMessages?(): Record<string, string>;
+  abstract scripts?(): string[];
+  abstract styles?(): string[];
+  public render(): WebView {
+    const policies = [
+      `default-src 'none';`,
+      `font-src ${this.panel.webview.cspSource} data:;`,
+      `img-src ${this.panel.webview.cspSource} data:;`,
+      `script-src ${this.panel.webview.cspSource};`,
+      `style-src ${this.panel.webview.cspSource};`,
+    ];
+    const styles = (this?.styles() || [])
+      .map(
+        (style) =>
+          `<link rel="stylesheet" href="${this.webviewUri(
+            this.extensionUri,
+            style,
+          )}">`,
+      )
+      .join("");
+    const scripts = (this?.scripts() || [])
+      .map(
+        (script) =>
+          `<script type="module" src="${this.webviewUri(
+            this.extensionUri,
+            script,
+          )}"></script>`,
+      )
+      .join("");
+
+    this.panel.webview.html = `<!DOCTYPE html>
+      <html lang="en">
+        <head>
+          <meta charset="UTF-8" />
+          <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+          <meta http-equiv="Content-Security-Policy" content="${policies.join(
+            " ",
+          )}" />
+          ${styles}
+          <title>${this.title}</title>
+        </head>
+        <body>
+          ${this.body()}
+          ${scripts}
+          ${
+            this?.l10nMessages
+              ? `<script type="application/json" id="l10n-messages">${JSON.stringify(this?.l10nMessages())}</script>`
```

Comment:
nit: it looks like only used as a string. How about make it in a data-* attribute to be clear it's not something executable?